pkg/controllers/resources/ingresses/syncer.go

```diff
@@ -29,43 +33,39 @@ func (s *ingressSyncer) SyncToHost(ctx *synccontext.SyncContext, vObj client.Obj
 	return s.SyncToHostCreate(ctx, vObj, s.translate(ctx.Context, vObj.(*networkingv1.Ingress)))
 }
 
-func (s *ingressSyncer) Sync(ctx *synccontext.SyncContext, pObj client.Object, vObj client.Object) (ctrl.Result, error) {
-	vIngress := vObj.(*networkingv1.Ingress)
-	pIngress := pObj.(*networkingv1.Ingress)
-
-	updated := s.translateUpdateBackwards(pObj.(*networkingv1.Ingress), vObj.(*networkingv1.Ingress))
-	if updated != nil {
-		ctx.Log.Infof("update virtual ingress %s/%s, because ingress class name is out of sync", vIngress.Namespace, vIngress.Name)
-		translator.PrintChanges(vIngress, updated, ctx.Log)
-		err := ctx.VirtualClient.Update(ctx.Context, updated)
-		if err != nil {
-			return ctrl.Result{}, err
+func (s *ingressSyncer) Sync(ctx *synccontext.SyncContext, pObj client.Object, vObj client.Object) (_ ctrl.Result, retErr error) {
+	patch, err := patcher.NewSyncerPatcher(ctx, pObj, vObj)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("new syncer patcher: %w", err)
+	}
+
+	defer func() {
+		if err := patch.Patch(ctx, pObj, vObj); err != nil {
+			retErr = utilerrors.NewAggregate([]error{retErr, err})
 		}
+	}()
 
-		// we will requeue anyways
-		return ctrl.Result{}, nil
+	pIngress, vIngress, _, _ := synccontext.Cast[*networkingv1.Ingress](ctx, pObj, vObj)
+	if vIngress.Spec.IngressClassName == nil && pIngress.Spec.IngressClassName != nil {
+		vIngress.Spec.IngressClassName = pIngress.Spec.IngressClassName
 	}
 
 	if !equality.Semantic.DeepEqual(vIngress.Status, pIngress.Status) {
 		newIngress := vIngress.DeepCopy()
 		newIngress.Status = pIngress.Status
-		ctx.Log.Infof("update virtual ingress %s/%s, because status is out of sync", vIngress.Namespace, vIngress.Name)
-		translator.PrintChanges(vIngress, newIngress, ctx.Log)
-		err := ctx.VirtualClient.Status().Update(ctx.Context, newIngress)
-		if err != nil {
-			return ctrl.Result{}, err
-		}
-
-		// we will requeue anyways
-		return ctrl.Result{}, nil
+		vObj = newIngress
 	}
+	annotations := maps.Clone(vObj.GetAnnotations())
 
 	newIngress := s.translateUpdate(ctx.Context, pIngress, vIngress)
```

Comment:
Why do we need to return newIngress here? Can we not directly change vIngress / pIngress?